File: Backend/controller/notice.js
import Notice from '../model/notice.js';
import asyncHandler from '../middlewares/async.js';
import ErrorResponse from '../utils/errorResponse.js';
import User from '../model/user.js'
export const addNewNotice = asyncHandler(async (req, res, next) => {
    try {
        // Create a new Notice instance with all data from req.body
        const { title, category, postedAt, expirationDate, startDate, author, content, targetCampuses, priorityLevel, status, visibility, relatedNotices, tags, approvalStatus, authorContact, lastModified } = req.body;

        // Create a new Notice instance
        const newNotice = new Notice({
            title,
            category,
            postedAt,
            expirationDate,
            startDate,
            author,
            content,
            targetCampuses,
            priorityLevel,
            status,
            visibility,
            relatedNotices,
            tags,
            approvalStatus,
            authorContact,
            lastModified,
        });

        // Save the new Notice to the database
        const savedNotice = await newNotice.save();

        res.status(200).json({
            success: true,
            data: savedNotice,
        });
    } catch (error) {
        // Handle any errors that occur during the save process
        console.error('Error saving Notice:', error);
        res.status(500).json({
            success: false,
            error: 'Internal Server Error',
        });
    }
});


export const getAllNotices = asyncHandler(async (req, res, next) => {
    const Notices = await Notice.find(); // Fetch all Notices from MongoDB
    res.status(200).json({
        success: true,
        notices: Notices
    });
});


// @desc Get Bookmarks
// @route GET/api/v1/Notice/get-bookmarked-Notices:id
// @access Private
export const getBookmarkedNotices = asyncHandler(async (req, res, next) => {
    const userId = req.params.userId; // Get userId from the route parameter
    // Populate the 'bookmarkedNotices' field with actual Notice documents.
    // This allows us to retrieve the complete Notice details associated with the user.
    const user = await User.findById(userId).populate({
        path: 'bookmarkedNotices',
        model: Notice
    });

    if (!user) {
        return next(new ErrorResponse(`User with id ${userId} not found`, 404));
    }

    const bookmarkedNotices = user.bookmarkedNotices.map(Notice => Notice.toObject());

    console.log(bookmarkedNotices);
    res.status(200).json({
        success: true,
        Notices: bookmarkedNotices
    });
});

// @desc Get A Notice By It's Id
// @route GET/api/v1/Notice/get-Notice-by-id:noticeId
// @access Private
export const getNoticeById = asyncHandler(async (req, res, next) => {
    const noticeId = req.params.noticeId;

    const notice = await Notice.findById(noticeId);

    if (!notice) {
        return next(new ErrorResponse(`Notice with id ${noticeId} not found`, 404));
    }


    console.log(notice);
    res.status(200).json({
        success: true,
        notice: notice
    });
});



// @desc Update a Notice
// @route PUT /api/v1/notice/:id
// @access Private

export const updateNotice = asyncHandler(async (req, res, next) => {
    let Notice = await Notice.findById(req.params.id);
    if (!Notice) {
        return next(new ErrorResponse(`No Notice with id of ${req.params.id}`, 404))
    }


    // //bootcamp can updated by only owner
    // if (course?.user !== req.user.id && req.user.role !== 'admin') {
    //     return next(new ErrorResponse(`User ${req.user.id} is not authorized to update Notice ${course._id}`, 401));
    // }
    Notice = await Notice.updateOne({ _id: req.params.id }, req.body, {
        new: true,
        runValidators: true
    })
    res.status(200).json({
        success: true,
        data: Notice
    })
});

File: lib/models/notice.dart
// ignore_for_file: public_member_api_docs, sort_constructors_first
import 'dart:convert';

class Notice {
  final String id;
  final String title; //
  final String category; //
  final DateTime postedAt;
  final DateTime expirationDate; //
  final DateTime startDate; //
  final String author; //
  final String content; //
  final List<String> targetCampuses; //
  final String priorityLevel; //
  final String status; //
  final String visibility; //
  final List<String> relatedNotices; //
  final List<String> tags; //
  final bool approvalStatus;
  final String authorContact;
  final DateTime lastModified;
  Notice({
    required this.id,
    required this.title,
    required this.category,
    required this.postedAt,
    required this.expirationDate,
    required this.startDate,
    required this.author,
    required this.content,
    required this.targetCampuses,
    required this.priorityLevel,
    required this.status,
    required this.visibility,
    required this.relatedNotices,
    required this.tags,
    required this.approvalStatus,
    required this.authorContact,
    required this.lastModified,
  });

  Notice copyWith({
    String? id,
    String? title,
    String? category,
    DateTime? postedAt,
    DateTime? expirationDate,
    DateTime? startDate,
    String? author,
    String? content,
    List<String>? targetCampuses,
    String? priorityLevel,
    String? status,
    String? visibility,
    List<String>? relatedNotices,
    List<String>? tags,
    bool? approvalStatus,
    String? authorContact,
    DateTime? lastModified,
  }) {
    return Notice(
      id: id ?? this.id,
      title: title ?? this.title,
      category: category ?? this.category,
      postedAt: postedAt ?? this.postedAt,
      expirationDate: expirationDate ?? this.expirationDate,
      startDate: startDate ?? this.startDate,
      author: author ?? this.author,
      content: content ?? this.content,
      targetCampuses: targetCampuses ?? this.targetCampuses,
      priorityLevel: priorityLevel ?? this.priorityLevel,
      status: status ?? this.status,
      visibility: visibility ?? this.visibility,
      relatedNotices: relatedNotices ?? this.relatedNotices,
      tags: tags ?? this.tags,
      approvalStatus: approvalStatus ?? this.approvalStatus,
      authorContact: authorContact ?? this.authorContact,
      lastModified: lastModified ?? this.lastModified,
    );
  }

  Map<String, dynamic> toMap() {
    return <String, dynamic>{
      '_id': id,
      'title': title,
      'category': category,
      'postedAt': postedAt.millisecondsSinceEpoch,
      'expirationDate': expirationDate.millisecondsSinceEpoch,
      'startDate': startDate.millisecondsSinceEpoch,
      'author': author,
      'content': content,
      'targetCampuses': targetCampuses,
      'priorityLevel': priorityLevel,
      'status': status,
      'visibility': visibility,
      'relatedNotices': relatedNotices,
      'tags': tags,
      'approvalStatus': approvalStatus,
      'authorContact': authorContact,
      'lastModified': lastModified.millisecondsSinceEpoch,
    };
  }

  factory Notice.fromMap(Map<String, dynamic> map) {
    return Notice(
      id: map['_id'] ?? '',
      title: map['title'] ?? '',
      category: map['category'] ?? '',
      postedAt: map['postedAt'] != null
          ? DateTime.fromMillisecondsSinceEpoch(
              map['postedAt'],
            )
          : DateTime.now(),
      expirationDate: map['expirationDate'] != null
          ? DateTime.fromMillisecondsSinceEpoch(
              map['expirationDate'],
            )
          : DateTime.now(),
      startDate: map['startDate'] != null
          ? DateTime.fromMillisecondsSinceEpoch(
              map['startDate'],
            )
          : DateTime.now(),
      author: map['author'] ?? '',
      content: map['content'] ?? '',
      targetCampuses: map['targetCampuses'] != null
          ? List<String>.from(
              map['targetCampuses'],
            )
          : [],
      priorityLevel: map['priorityLevel'] ?? '',
      status: map['status'] ?? '',
      visibility: map['visibility'] ?? '',
      relatedNotices: map['relatedNotices'] != null
          ? List<String>.from(
              map['relatedNotices'],
            )
          : [],
      tags: map['tags'] != null
          ? List<String>.from(
              map['tags'],
            )
          : [],
      approvalStatus: map['approvalStatus'] ?? false,
      authorContact: map['authorContact'] ?? '',
      lastModified: map['lastModified'] != null
          ? DateTime.fromMillisecondsSinceEpoch(
              map['lastModified'],
            )
          : DateTime.now(),
    );
  }

  String toJson() => json.encode(toMap());

  factory Notice.fromJson(String source) =>
      Notice.fromMap(json.decode(source) as Map<String, dynamic>);
}
